packages/flutter/lib/src/rendering/mouse_cursor.dart

```diff
@@ -0,0 +1,448 @@
+        return annotation;
+      }
+    }
+    return const _FallbackAnnotation();
+  }
+
+  // Handles device update and changes mouse cursors.
+  void _handleDeviceUpdateMouseCursor(MouseTrackerUpdateDetails details) {
+    final int device = details.device;
+
+    if (details.triggeringEvent is PointerRemovedEvent) {
+      _mouseCursorStates.remove(device);
+      return;
+    }
+
+    bool debugHadState;
+    assert(() {
+      debugHadState = _mouseCursorStates.containsKey(device);
+      return true;
+    }());
+    final _MouseCursorState state = _mouseCursorStates.putIfAbsent(device, () => _MouseCursorState());
+
+    final MouseCursorSession lastSession = state.session;
+    assert(lastSession != null || !debugHadState);
+    final MouseTrackerAnnotation nextAnnotation = _findCursorAnnotation(details.nextAnnotations);
+    if (lastSession?.cursor == nextAnnotation.cursor)
+      return;
+
+    final MouseCursorSession nextSession = nextAnnotation.cursor.createSession(device);
+    state.session = nextSession;
+
+    lastSession?.dispose();
+    nextSession.activate();
+  }
+}
+
+/// Manages the duration that a pointing device should display a specific mouse
+/// cursor.
+///
+/// While [MouseCursor] classes describe the kind of cursors, [MouseCursorSession]
+/// classes represents a continuous use of the cursor on a pointing device. The
+/// [MouseCursorSession] classes can be stateful. For example, a cursor that
+/// needs to load resources might want to set a temporary cursor first, then
+/// switch to the correct cursor after the load is completed.
+///
+/// A [MouseCursorSession] have the following lifecycle:
+/// 
+///  * When a pointing device should start displaying a cursor, [MouseTracker]
+///    creates a session by calling [MouseCursor.createSession] on the target
+///    cursor, and stores it in a table associated with the device.
+///  * [MouseTracker] then immediately calls the session's [activate], where the
+///    session should fetch resources and make system calls.
+///  * When the pointing device should start displaying a different cursor, which
+///    is determined by [MouseCursor]'s `==`, [MouseTracker] calls [dispose] on the
+///    session, and this session will no longer be used in the future.
+abstract class MouseCursorSession {
+  /// Create a session.
+  ///
+  /// All arguments must be non-null.
+  MouseCursorSession(this.cursor, this.device)
+    : assert(cursor != null),
+      assert(device != null);
+
+  /// The cursor that created this session.
+  final MouseCursor cursor;
+
+  /// The device ID of the pointing device.
+  final int device;
+
+  /// Override this method to do the work of changing the cursor of the device.
+  ///
+  /// Called right after this session is created.
+  ///
+  /// This method should collect necessary resources, then make system calls
+  /// (usually through [SystemChannels.mouseCursor]) to apply the effect.
+  /// This method may change the state of this class.
```

Comment:
I changed it to 
```dart
  /// This method has full control over the cursor until the [dispose] call, and
  /// can make system calls to change the pointer cursor as many times as
  /// necessary (usually through [SystemChannels.mouseCursor]). It can also
  /// collect resources, and store the result in this object.
```